### ISSUE TITLE
Assignment without let

### ISSUE THREAD

**Issue by Timmmm** [rems-project/sail#201]
(Sorry for the barrage of questions!)

Suppose you have this code:

```
// In one file.
register registers_written : bool
```

```
// In another file far far away.
function ext_pre_step_hook()  -> unit = {
    registers_written = false;
}
```

If you later rename `registers_written` then you will get a silent bug, because the `registers_written = false;` now creates a local mutable variable (I think?).

This is the same mistake Python (and some other languages) made by not having a distinct syntax for "update an existing variable" or "create a new variable". Maybe worth reconsidering depending on how frozen you consider the language?

**Comment by martinberger** [comment]
I agree, I think this is a language bug. 

This caused hard-to-debug errors at my previous employer. I suspect the language design here is a legacy issue, since Sail was intended to be as close to ASL as possible? If this cannot be changed for compatibility reasons, then I recommend having the option to emit compiler warnings in all such situations.

**Comment by arichardson** [comment]
See https://github.com/rems-project/sail/issues/191. Rust and OCaml apparently have the same "bug", so it's not unexpected.

**Comment by Alasdair** [comment]
Yes, it was for consistency with ASL originally and I agree it was a mistake. We added the `var` keyword so you can be explicit:

```
// In another file far far away.
function ext_pre_step_hook()  -> unit = {
    var registers_written = false;
}
```

but we never made it mandatory. It would be good to make additional steps towards that, and adding a warning and updating the RISC-V spec would be a logical next step.

**Comment by Alasdair** [comment]
In the next Sail version I can add a `-strict_assignments` flag or similar, and a warning. Then in the subsequent version we could consider making it the default.

**Comment by martinberger** [comment]
> *In the next Sail version I can add ...*

Is there a consistent way that the Sail maintainers let the RISCV Sail model maintainers know what is in the latest and planned releases of Sail, so the latter can plan and use the new features? Or by osmosis? @Alasdair

**Comment by Alasdair** [comment]
Well hopefully me attending the RISC-V formal meetings on tuesday can be useful for that purpose!

The `var` syntax has been around for a while,  because we have a re-write that introduces it automatically for targets that require explicit declarations. Making it required has just been something on the very back of my mind for a long time. It wasn't until this issue today that I put a bit more thought into seriously making it the default.

**Comment by Timmmm** [comment]
@Alasdair did `-strict_assignments` ever get added? This is actually worse than I thought. Because Sail's compilation model is just to concatenate all the inputs, if you have a load of files (like e.g. the RISC-V Sail model) it can be hard to keep track of the order of things.

I think that's already a bit of an issue with scattered functions/mappings, but it also interacts really badly with this bug. If you write this:

```
// a.sail
register foo : bool

// b.sail
function bar() = {
  foo = true;
  ...
}
```

That works fine, but suppose you didn't realise that `b.sail` actually comes before `a.sail`. Now you actually get this:

```
function bar() = {
  foo = true;
  ...
}

register foo : bool
```

Which silently changes the behaviour of the code so `foo` is now a local variable. I only eventually realised this by reading the generated C code. :-(

I'd be *really* grateful if this could be fixed! :-)

**Comment by Timmmm** [comment]
This is now available as `--strict-var`. I'll leave this open so we remember to eventually make it the default.
